llama.cpp

```diff
@@ -9470,6 +9715,35 @@ struct llm_build_context {
         return gf;
     }
 
+    static ggml_tensor * lora_mul_mat(
+        llama_context & lctx, 
+        ggml_context * ctx0, 
+        ggml_tensor * weight, 
+        ggml_tensor * cur) {
+    ggml_tensor * mm = ggml_mul_mat(ctx0, weight, cur);
+
+    auto it = lctx.lora_weights_map.find(weight->name);
+    if (it == lctx.lora_weights_map.end()) {
+        return mm;
+    }
+
+    ggml_tensor * loraA = it->second.loraA;
+    ggml_tensor * loraB = it->second.loraB;
+
+    ggml_tensor * t_lora = ggml_mul_mat(ctx0,
+                ggml_mul_mat(ctx0, loraA, loraB), 
+                cur
+            );
```

Comment:
This matrix multiplication is why it is extremely slow. It materialized a matrix of the same dimension as the weight, and then multiplies that matrix with the hidden state. It is essentially doubling the number of matrix multiplications. The `lora_mul_mat` that I gave you does something very different:
```C++
    ggml_tensor * t_lora =
        ggml_mul_mat(ctx0,
            loraB,
            ggml_mul_mat(ctx0,
                loraA,
                cur
            )
        );
```
These matrix multiplications produce only much smaller vectors as the result. This would be a lot faster, but to make this work `loraA` will need to be transposed. For optimal efficiency, new kernels optimized for the sizes of the lora matrices (very low number of columns) will also need to be implemented. Properly implemented, this will only add a small overhead to the computation (I estimate 3-10%, depending on the rank of the lora).

Reply:
@slaren Ops! Thanks! Fixed and it's faster (to the point that I do not perceive a difference from not using the lora).

Is this closed PR #996 something could work to optimise performance of the lora matrices?

Anyway, I wanted to first set up the current PR so that one can spin-up a server and hot-swap adapters. Do you think I should focus on optimising thin matrices multiplication first instead?

Reply:
@ltoniazzi Great to hear that it works on your side.

I think this PR can be a very good starting point for demo'ing that it works. However, I'll implement proper API for it in #8332 

In the meantime, it would be useful to a look on how to convert HF "safetensors" adapter to gguf. Potentially we can introduce `--lora` param to `convert_hf_to_gguf.py` which already had all the logics for tensor name. An example for the safetensors adapter can be found here: https://huggingface.co/grimjim/Llama-3-Instruct-abliteration-LoRA-8B/tree/main

Reply:
@ngxson Thanks for the comments 😊 !! Btw, I think in your version the large BA matrix is coming up[ here](https://github.com/ngxson/llama.cpp/compare/master...ngxson:llama.cpp:4e28ad40a099c7f618abf8ae113c4e56ee7705e8#diff-57725fa4c99fac28d0283eaa8b59c5e030355af0c8fbaf33145a64e3f2aa3420R18500).

Yes, let me know if you need help. I can start to have a look at converting loras from safetensors to gguf. (This conversion was discussed [here](https://github.com/ggerganov/llama.cpp/discussions/7850#discussioncomment-9729938) as having a script `convert-lora-to-ggml.py` before it was removed in #7204 as it might have to be maintained for each model architecture.)

Btw, how do you render the model's graph?

Reply:
> Btw, how do you render the model's graph?

You can do `ggml_graph_dump_dot(gf, NULL, "/tmp/graph.dot")`

Then using `dot -Tsvg /tmp/graph.dot -o /tmp/graph.svg` to convert it to svg (default command converts it to png, but I find it's quite difficult to open a big graph). You can also use online service: https://dreampuf.github.io/GraphvizOnline/

A small trick is to change `rankdir` to `TB`, it fits better on screen. And finally this won't work if graph is too big. I can only print graph of the tinyllama `stories15M.gguf` model

Reply:
You can also limit the loop in `build_llama` to include only one layer in the graph.